apps/argos_aggregation/lib/argos_aggregation/elastic_serach/indexer.ex

```diff
@@ -32,6 +32,9 @@ defmodule ArgosAggregation.ElasticSearch.Indexer do
     end
 
   end
+  def index({:ok, %{"core_fields" => _} = core }) do
```

Comment:
Wozu braucht es das Patternmatching auf "core_fields"? Könntest du den thesauri Harvester auch so schreiben, dass er einfach eine Map übergibt (analog zu den restlichen Harvestern), anstatt {:ok, map}?

Ich hatte hier auch mal eine Variante mit {:ok, map}, allerdings war dann meine Überlegung folgende: Wenn ein Harvester (oder irgendein anderer Teil der Anwendung) etwas indizieren möchte, dann sollte es in dessen Aufgabenbereich fallen alle vorherigen Schritte zu evaluieren. Am Ende gibt der dann die Map, oder wenn es auf seiner Seite ein {:error, ..} o.ä. gab wird die index() Funktion eben erst gar nicht aufgerufen.

Reply:
Der Match auf core_fields wird implizit durch die validate Funktion, die gleich als erstes Aufgerufen wird ohnehin vollzogen, ich habe das durch das Pattern nur explizit gemacht. Ich hatte in den Tests den Fall, das falsche Maps übergeben wurden, die dann erst im validate auf keine passende Funktion gestossen sind und dann einen Non Matching Function Error verursacht haben.

Reply:
> analog zu den restlichen Harvestern

Die Datenprovider, die mehr als ein Dokument zurück liefern, erzeugen zur Zeit alle eine Liste mit [ {:ok, map}, {:ok, map} ]  die direkt in den Indexer gefüttert wird. An das Schema hält sich auch der Thesaurus Harvester

Reply:
> Ich hatte hier auch mal eine Variante mit {:ok, map}, allerdings war dann meine Überlegung folgende: Wenn ein Harvester (oder irgendein anderer Teil der Anwendung) etwas indizieren möchte, dann sollte es in dessen Aufgabenbereich fallen alle vorherigen Schritte zu evaluieren. Am Ende gibt der dann die Map, oder wenn es auf seiner Seite ein {:error, ..} o.ä. gab wird die index() Funktion eben erst gar nicht aufgerufen.

Können wir bitte versuchen, solche Schnittstellen-Spezifikation vorab zu definieren und am Besten im Code zu enforcen. Das immer erst im Nachhinein zu klären ist etwas suboptimal. 
Gut wäre eine einheitliche Definition von Übergabe- und Rückgabewerten bei jeder Schnittstelle, die von verschieden Leuten entwickelt wird.

Reply:
Es war ja im Code enforced oder nicht?

Reply:
> > analog zu den restlichen Harvestern
> 
> Die Datenprovider, die mehr als ein Dokument zurück liefern, erzeugen zur Zeit alle eine Liste mit [ {:ok, map}, {:ok, map} ] die direkt in den Indexer gefüttert wird. An das Schema hält sich auch der Thesaurus Harvester
> 
> Ich habe jetzt gesehen, dass dein Bibliography Harvester das nicht macht. Daher siehe unten.

Ja, die Dataprovider schon, aber ich glaube die Harvester nicht (die sind ja dann diejenigen, die die Indizierung anstoßen.

Reply:
> Der Match auf core_fields wird implizit durch die validate Funktion, die gleich als erstes aufgerufen wird, ohnehin vollzogen, ich habe das durch das Pattern nur explizit gemacht. Ich hatte in den Tests den Fall, das falsche Maps übergeben wurden, die dann erst im validate auf keine passende Funktion gestossen sind und dann einen Non Matching Function Error verursacht haben.

Dann würde ich folgendes vorschlagen: Anstatt hier auf core_fields zu matchen, würde ich eher bei der validate ansetzen und die ein {:error, :unknown_ecto_schema} (o.ä., kommt drauf an was bei deinem konkreten Fall das Problem war) zurückgeben lassen.

Im Grunde nimmt der match an der Stelle validation vorweg, ich fände es besser es bei der restlichen validation zu halten. Was ist z.B. wenn wir mal was ohne "core_fields" indizieren wollen? Mir ist nicht ganz klar was das vorziehen bringt, außer dass jetzt an zwei Stellen evaluiert wird.

Reply:
Das Vorziehen macht den konkreten Kontrakt nach außen hin sichtbar. Es sagt explizit: "Ich erwarte core-fields", was es defacto ja auch tut nur eben bislang unter versteckt. Wenn man irgendwann etwas ohne core-fields indizieren wollen würde, müsste bisher den validator anpassen oder jetzt, eine neue Funktionssignatur anlegen, der Aufwand sollte derselbe sein. Letzteres fühlt sich für mich mehr nach Elixir an. Zumindest verstehe ich das Konzept so, dass man eher je neue Anforderung einen neue Funktion schafft, als eine bestehende weiter zu verästeln. 
Das Pattern nimmt auch nicht die Validierung weg, da es ja nicht auf die exakte Entität matcht. Obwohl man das natürlich gut machen könnte, dann hätte man halt fünf explizite index Signaturen, fände ich persönlich sogar noch besser.

Reply:
> > > analog zu den restlichen Harvestern
> > 
> > 
> > Die Datenprovider, die mehr als ein Dokument zurück liefern, erzeugen zur Zeit alle eine Liste mit [ {:ok, map}, {:ok, map} ] die direkt in den Indexer gefüttert wird. An das Schema hält sich auch der Thesaurus Harvester
> > Ich habe jetzt gesehen, dass dein Bibliography Harvester das nicht macht. Daher siehe unten.
> 
> Ja, die Dataprovider schon, aber ich glaube die Harvester nicht (die sind ja dann diejenigen, die die Indizierung anstoßen).

Das war was ich mit Schnittstellen meinte, das war bislang nicht einheitlich geregelt. Gazetteer hat die Tuple einfach weitergereicht, daran hatte ich mich orientiert. Ich habe das aber mittlerweile für beide angepasst.

Reply:
> Das Pattern nimmt auch nicht die Validierung weg, da es ja nicht auf die exakte Entität matcht. Obwohl man das natürlich gut machen könnte, dann hätte man halt fünf explizite index Signaturen, fände ich persönlich sogar noch besser.

Genau. Vielleicht habe ich mich etwas blöde ausgedrückt: Das ist auch ein super kleiner Punkt, aber ich würde die Sachen lieber direkt an einer Stelle zusammenhalten bevor es ggf. irgendwann unübersichtlich wird. Deshalb fände ich es schöner, wenn man sukzessive weitere Signaturen von validate() hinzufügt, anstatt schon vor dem Aufruf validate() Parameter zurückzuweisen.

Dann bleibt es bei exakt einer Funktion (mit mehreren Signaturen), die kontrolliert was überhaupt indiziert werden darf.

Reply:
Um hier voran zu kommen, habe ich jetzt die index function wieder zurück geändert.